src/crystal/system/print_error.cr

```diff
@@ -1,15 +1,150 @@
 module Crystal::System
-  # Prints directly to stderr without going through an IO.
+  # Prints directly to stderr without going through an `IO::FileDescriptor`.
   # This is useful for error messages from components that are required for
   # IO to work (fibers, scheduler, event_loop).
   def self.print_error(message, *args)
-    {% if flag?(:unix) %}
-      LibC.dprintf 2, message, *args
-    {% elsif flag?(:win32) %}
-      buffer = StaticArray(UInt8, 512).new(0_u8)
-      len = LibC.snprintf(buffer, buffer.size, message, *args)
-      LibC._write 2, buffer, len
-    {% end %}
+    print_error(message, *args) do |bytes|
+      {% if flag?(:unix) || flag?(:wasm32) %}
+        LibC.write 2, bytes, bytes.size
+      {% elsif flag?(:win32) %}
+        LibC._write 2, bytes, bytes.size
+      {% end %}
+    end
+  end
+
+  # Minimal drop-in replacement for a C `printf` function. Yields successive
+  # `Bytes` to the block, which should do the actual printing.
+  #
+  # *format* only supports the `%(l|ll)?[dpsux]` format specifiers; more should
+  # be added only when we need them or an external library passes its own format
+  # string to here. *format* must also be null-terminated.
+  #
+  # Since this method may be called under low memory conditions or even with a
+  # corrupted heap, its implementation should be as low-level as possible,
+  # avoiding memory allocations.
+  #
+  # NOTE: C's `printf` is incompatible with Crystal's `sprintf`, because the
+  # latter does not support argument width specifiers nor `%p`.
+  def self.print_error(format, *args, &)
+    ptr, format_len = to_unsafe_string(format)
+    finish = ptr + format_len
+    arg_index = 0
+
+    while ptr < finish
+      next_percent = ptr
+      while next_percent < finish && !(next_percent.value === '%')
+        next_percent += 1
+      end
+      yield Slice.new(ptr, next_percent - ptr)
+
+      fmt_ptr = next_percent + 1
+      width = 0
+      if fmt_ptr.value === 'l'
+        width = 1
+        fmt_ptr += 1
+        if fmt_ptr.value === 'l'
+          width = 2
+          fmt_ptr += 1
+        end
+      end
+
+      break unless fmt_ptr < finish
+
+      case fmt_ptr.value
+      when 's'
+        if (arg = args[arg_index].as?(String | UInt8*)).nil?
+          yield "(???)".to_slice
+        else
+          yield Slice.new(*to_unsafe_string(arg))
+        end
```

Comment:
**nitpick(style):** I think code readability would improve if we swap branches for simplified, positive conditions.
```suggestion
        if arg = args[arg_index].as?(String | UInt8*)
          yield Slice.new(*to_unsafe_string(arg))
        else
          yield "(???)".to_slice
        end
```
Ditto all below.